src/main/kotlin/com/manimdsl/ManimParserVistor.kt

```diff
@@ -2,61 +2,140 @@ package com.manimdsl
+            visit(ctx.type()) as Type
+        } else {
+            rhsType
+        }
+
+        semanticAnalyser.failIfIncompatibleTypes(lhsType, rhsType, ctx.expr().text, ctx)
+
+        currentSymbolTable.addVariable(identifier, rhsType)
+        return DeclarationNode(ctx.start.line, identifier, rhs)
     }
 
-    override fun visitDeclarationStatement(ctx: ManimParser.DeclarationStatementContext?): ASTNode {
-        return super.visitDeclarationStatement(ctx)
+    override fun visitAssignmentStatement(ctx: ManimParser.AssignmentStatementContext): AssignmentNode {
+        val expression = visit(ctx.expr()) as ExpressionNode
+        val identifier = ctx.IDENT().symbol.text
+        val rhsType = semanticAnalyser.inferType(currentSymbolTable, expression)
+        val identifierType = currentSymbolTable.getTypeOf(identifier)
+
+        semanticAnalyser.undeclaredIdentifier(currentSymbolTable, identifier, ctx)
+        semanticAnalyser.failIfIncompatibleTypes(identifierType, rhsType, ctx.expr().text, ctx)
+        return AssignmentNode(ctx.start.line, identifier, expression)
     }
 
-    override fun visitAssignmentStatement(ctx: ManimParser.AssignmentStatementContext?): ASTNode {
-        return super.visitAssignmentStatement(ctx)
+    override fun visitMethodCallStatement(ctx: ManimParser.MethodCallStatementContext): MethodCallNode {
+        return visitMethodCall(ctx.method_call() as ManimParser.MethodCallContext)
     }
 
-    override fun visitMethodCallStatement(ctx: ManimParser.MethodCallStatementContext?): ASTNode {
-        return super.visitMethodCallStatement(ctx)
+    override fun visitMethodCallExpression(ctx: ManimParser.MethodCallExpressionContext): MethodCallNode {
+        return visitMethodCall(ctx.method_call() as ManimParser.MethodCallContext)
     }
 
-    override fun visitArg_list(ctx: ManimParser.Arg_listContext?): ASTNode {
-        return super.visitArg_list(ctx)
+    override fun visitArgumentList(ctx: ManimParser.ArgumentListContext?): ArgumentNode {
+        return ArgumentNode((ctx?.expr()
+                ?: listOf<ManimParser.ExprContext>()).map { visit(it) as ExpressionNode })
     }
 
-    override fun visitMethodCall(ctx: ManimParser.MethodCallContext?): ASTNode {
-        return super.visitMethodCall(ctx)
+    override fun visitMethodCall(ctx: ManimParser.MethodCallContext): MethodCallNode {
+        // Type signature of methods to be determined by symbol table
+        val arguments: List<ExpressionNode> = visitArgumentList(ctx.arg_list() as ManimParser.ArgumentListContext?).arguments
+        val identifier = ctx.IDENT(0).symbol.text
+        val methodName = ctx.IDENT(1).symbol.text
+
+        semanticAnalyser.undeclaredIdentifier(currentSymbolTable, identifier, ctx)
+        semanticAnalyser.failIfNotDataStructure(currentSymbolTable, identifier, ctx)
+        semanticAnalyser.notValidMethodNameForDataStructure(currentSymbolTable, identifier, methodName, ctx)
```

Comment:
Rename above to sound more conditional and not absolute e.g. undeclaredIdentifier -> undeclaredIdentifierCheck